controller/src/main/java/io/pravega/controller/server/SegmentHelper.java

```diff
@@ -1035,6 +1044,199 @@ public void authTokenCheckFailed(WireCommands.AuthTokenCheckFailed authTokenChec
         return result;
     }
 
+    /**
+     * The method sends a WireCommand to iterate over table keys.
+     * @param scope Stream scope.
+     * @param stream Stream name.
+     * @param suggestedKeyCount Suggested number of {@link TableKey}s to be returned by the SegmentStore.
+     * @param state Last known state of the iterator.
+     * @param hostControllerStore Host Controller store.
+     * @param clientCF Client connection factory.
+     * @param delegationToken The token to be presented to the segmentstore.
+     * @param clientRequestId Request id.
+     * @return An {@link AsyncIterator} that will return the next set of {@link TableKey}s returned from the SegmentStore.
+     */
+    public AsyncIterator<TableSegment.IteratorItem<TableKey<byte[]>>> readTableKeys(final String scope,
+                                                                                    final String stream,
+                                                                                    final int suggestedKeyCount,
+                                                                                    final IteratorState state,
+                                                                                    final HostControllerStore hostControllerStore,
+                                                                                    final ConnectionFactory clientCF,
+                                                                                    final String delegationToken,
+                                                                                    final long clientRequestId) {
+
+        final Controller.NodeUri uri = getSegmentUri(scope, stream, 0L, hostControllerStore);
+        final String qualifiedName = getScopedStreamName(scope, stream);
+        final WireCommandType type = WireCommandType.READ_TABLE_KEYS;
+        final long requestId = (clientRequestId == RequestTag.NON_EXISTENT_ID) ? idGenerator.get() : clientRequestId;
+
+        return new ContinuationTokenAsyncIterator<>(token -> {
```

Comment:
You have a lambda that contains an anonymous implementation of an interface. Let's refactor this code and make this a separate method, and the `FailingReplyProcessor` its own sub-class. That will make reading this much easier.